// @cooklang/cooklang-ts/dist/Parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parser = void 0;
const tokens_1 = require("./tokens");
class Parser {
    parse(source) {
        const metadata = {};
        const steps = [];
        const shoppingList = {};
        source = source
            .replace(tokens_1.comment, '')
            .replace(tokens_1.blockComment, ' ');
        for (let match of source.matchAll(tokens_1.shoppingListToken)) {
            const groups = match.groups;
            if (!groups)
                continue;
            shoppingList[groups.name] = parseShoppingListItems(groups.items || '');
            source = source.substring(0, match.index || 0);
            +source.substring((match.index || 0) + match[0].length);
        }
        const lines = source
            .split('\n')
            .filter(l => l.trim().length > 0);
        for (let line of lines) {
            const step = [];
            let pos = 0;
            for (let match of line.matchAll(tokens_1.tokens)) {
                const groups = match.groups;
                if (!groups)
                    continue;
                if (pos < (match.index || 0)) {
                    step.push({
                        type: 'text',
                        value: line.substring(pos, match.index),
                    });
                }
                if (groups.key && groups.value) {
                    metadata[groups.key.trim()] = groups.value.trim();
                }
                if (groups.sIngredientName) {
                    step.push({
                        type: 'ingredient',
                        name: groups.sIngredientName,
                        quantity: 'some',
                    });
                }
                if (groups.mIngredientName) {
                    step.push({
                        type: 'ingredient',
                        name: groups.mIngredientName,
                        quantity: parseQuantity(groups.mIngredientQuantity, 'some'),
                        units: parseUnits(groups.mIngredientUnits),
                    });
                }
                if (groups.sCookwareName) {
                    step.push({
                        type: 'cookware',
                        name: groups.sCookwareName,
                    });
                }
                if (groups.mCookwareName) {
                    step.push({
                        type: 'cookware',
                        name: groups.mCookwareName,
                        quantity: parseQuantity(groups.mCookwareQuantity),
                    });
                }
                if (groups.timerQuantity) {
                    step.push({
                        type: 'timer',
                        name: groups.timerName,
                        quantity: parseQuantity(groups.timerQuantity),
                        units: parseUnits(groups.timerUnits),
                    });
                }
                pos = (match.index || 0) + match[0].length;
            }
            if (pos < line.length) {
                step.push({
                    type: 'text',
                    value: line.substring(pos),
                });
            }
            if (step.length > 0)
                steps.push(step);
        }
        return { metadata, steps, shoppingList };
    }
}
exports.Parser = Parser;
function parseQuantity(quantity, defaultText) {
    if (!quantity || quantity.trim() == '') {
        if (defaultText)
            return defaultText;
        return undefined;
    }
    quantity = quantity.trim();
    const [left, right] = quantity.split('/');
    const [numLeft, numRight] = [Number(left), Number(right)];
    if (!isNaN(numLeft) && !numRight)
        return numLeft;
    else if (!isNaN(numLeft) && !isNaN(numRight) && !(left.startsWith('0') || right.startsWith('0')))
        return numLeft / numRight;
    return quantity.trim();
}
function parseUnits(units) {
    if (!units)
        return undefined;
    return units.trim();
}
function parseShoppingListItems(items) {
    const list = [];
    for (let item of items.split('\n')) {
        item = item.trim();
        if (item == '')
            continue;
        const [name, synonym] = item.split('|');
        list.push({
            name: name.trim(),
            synonym: synonym?.trim() || '',
        });
    }
    return list;
}
